### PR TITLE
change: Migrate workspace to pasta_curves-0.5

### PR DESCRIPTION
**Pull request by CPerezz**
This ports the the workspace to the `pasta_curves-0.5.0`, `ff-0.13` and `group-0.13` deps.

I took a chance in this PR also to update a bit the `halo2-gadgets` crate to be more in-line with upstream.
Resolves: #132

### CODE REVIEW

**Comment by CPerezz** [comment]
> There seem to be some type issues and lint failures ATM. I'll finish the review when those are fixed. Thanks a lot for making this change!! :)

I forgot to push the last commits!!!! :sob: It should be ready now! Apologies!!!

File: halo2_proofs/examples/shuffle.rs [hunk 1]
```diff
@@ -253,12 +251,12 @@ impl<F: FieldExt, const W: usize, const H: usize> Circuit<F> for MyCircuit<F, W,
     }
 }
 
-fn test_mock_prover<F: FieldExt, const W: usize, const H: usize>(
+fn test_mock_prover<F: Field + Ord + FromUniformBytes<64>, const W: usize, const H: usize>(
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply halo2_proofs/examples/shuffle.rs:254]: Yepp! Makes sense!

File: halo2_gadgets/src/utilities.rs [hunk 2]
```diff
@@ -167,7 +166,7 @@ pub fn bitrange_subset<F: PrimeFieldBits>(field_elem: &F, bitrange: Range<usize>
 
 /// Check that an expression is in the small range [0..range),
 /// i.e. 0 ≤ word < range.
-pub fn range_check<F: FieldExt>(word: Expression<F>, range: usize) -> Expression<F> {
+pub fn range_check<F: PrimeField>(word: Expression<F>, range: usize) -> Expression<F> {
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply halo2_gadgets/src/utilities.rs:169]: You're right! I missed it! Will try to fix all the occurrences!

File: halo2_gadgets/benches/poseidon.rs [hunk 3]
```diff
@@ -123,7 +123,7 @@ where
 #[derive(Debug, Clone, Copy)]
 struct MySpec<const WIDTH: usize, const RATE: usize>;
 
-impl<const WIDTH: usize, const RATE: usize> Spec<Fp, WIDTH, RATE> for MySpec<WIDTH, RATE> {
+impl Spec<Fp, 3, 2> for MySpec<3, 2> {
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply halo2_gadgets/benches/poseidon.rs:126]: From upstream main. But I looked at an old blame. So I fixed by porting to latest upstream.

The constants anyway are at the end of the file when you call the benchmark.

File: halo2_gadgets/src/poseidon/primitives.rs [hunk 4]
```diff
@@ -50,33 +52,41 @@ pub trait Spec<F: FieldExt, const T: usize, const RATE: usize>: fmt::Debug {
...
-                    *rc = value;
-                }
-                rc_row
-            })
-            .collect();
+    fn constants() -> (Vec<[F; T]>, Mds<F, T>, Mds<F, T>);
+}
+
+/// Generates `(round_constants, mds, mds^-1)` corresponding to this specification.
+pub fn generate_constants<
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply halo2_gadgets/src/poseidon/primitives.rs:59]: I just followed [upstream/main](https://github.com/zcash/halo2/blob/47f2cc83498de27bcf1641e9f2e02e0c86551c7e/halo2_gadgets/src/poseidon/primitives.rs#L56).

In any case, I agree is a bit weird.

File: halo2_gadgets/benches/poseidon.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (han0110) [reply halo2_gadgets/benches/poseidon.rs:126]: It seems the upstream is updated to have these removed in https://github.com/zcash/halo2/pull/712 last month, but I think it's fine to have it as is.

File: halo2_proofs/src/plonk.rs [hunk 5]
```diff
@@ -151,12 +150,18 @@ where
     pub fn from_bytes<ConcreteCircuit: Circuit<C::Scalar>>(
         mut bytes: &[u8],
         format: SerdeFormat,
-    ) -> io::Result<Self> {
+    ) -> io::Result<Self>
+    where
+        C::Scalar: FromUniformBytes<64>,
+    {
```
Comment (han0110) [inline halo2_proofs/src/plonk.rs:156]: Seems duplicated as line 62

File: halo2_proofs/src/plonk.rs [hunk 6]
```diff
@@ -330,7 +338,10 @@ where
     pub fn read<R: io::Read, ConcreteCircuit: Circuit<C::Scalar>>(
         reader: &mut R,
         format: SerdeFormat,
-    ) -> io::Result<Self> {
+    ) -> io::Result<Self>
+    where
+        C::Scalar: FromUniformBytes<64>,
+    {
```
Comment (han0110) [inline halo2_proofs/src/plonk.rs:344]: ditto (as line 281)

File: halo2_proofs/src/plonk.rs [hunk 7]
```diff
@@ -364,7 +375,10 @@ where
     pub fn from_bytes<ConcreteCircuit: Circuit<C::Scalar>>(
         mut bytes: &[u8],
         format: SerdeFormat,
-    ) -> io::Result<Self> {
+    ) -> io::Result<Self>
+    where
+        C::Scalar: FromUniformBytes<64>,
+    {
```
Comment (han0110) [inline halo2_proofs/src/plonk.rs:381]: ditto

**Review by han0110 — APPROVED** [review]
LGTM! Thanks for taking care of this!

File: halo2_gadgets/src/utilities.rs [hunk 8]
```diff
@@ -130,15 +129,15 @@ impl<F: Field> RangeConstrained<F, AssignedCell<F, F>> {
 }
 
 /// Checks that an expression is either 1 or 0.
-pub fn bool_check<F: FieldExt>(value: Expression<F>) -> Expression<F> {
+pub fn bool_check<F: PrimeField>(value: Expression<F>) -> Expression<F> {
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply halo2_gadgets/src/utilities.rs:132]: No as `range_check` requires `From<u64>` which is required via `PrimeField` upstream.

File: halo2_proofs/src/plonk.rs (hunk 5 above)
Comment (CPerezz) [reply halo2_proofs/src/plonk.rs:156]: WDYM by duplicated?? Not sure I follow

File: halo2_gadgets/benches/poseidon.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (CPerezz) [reply halo2_gadgets/benches/poseidon.rs:126]: I updated to the latest upstream version! :)

**Pull request merged**